draft-ietf-mops-streaming-opcons.md

```diff
@@ -341,6 +371,60 @@ We note that other operators saw similar spikes during this time period. Craig L
 - A 30% increase in upstream traffic over their pre-pandemic levels, and
 - A steady increase in the overall volume of DDoS traffic, with amounts exceeding the pre-pandemic levels by 40%. (He attributed this increase to the significant rise in gaming-related DDoS attacks ({{LabovitzDDoS}}), as gaming usage also increased.)
 
+# Latency Considerations
+
+Streaming media latency refers to the "glass-to-glass" time duration, which is the delay between the real-life occurence of a recorded event and the media being appropriately displayed on an end user's device.  Note that this is different from the network latency (defined as the time for a packet to cross a network from end to end) because it includes video encoding time, and for most cases also ingest to an intermediate service such as a CDN or other video distribution service, rather than a direct connection to an end user.
+
+Streaming media can be usefully categorized according to the application's latency requirements into a few rough categories:
+
+  * ultra low latency    (less than 1 second)
+  * low latency live     (less than 10 seconds)
+  * non-low-latency live (10 seconds to a few minutes)
+  * on-demand            (hours or more)
+
+## Ultra Low Latency
+
+Ultra low latency delivery of media is defined here as having a glass-to-glass delay target under one second.
+
+This level of latency is sometimes necessary for real-time interactive applications such as video conferencing, operation of remote control devices or vehicles, or remotely hosted real-time gaming systems.  Some media content providers aim to achieve this level of latency for live media events involving sports, but have usually so far been unsuccessful over the internet at scale, though it is often possible within a localized environment with a controlled network, such as inside a specific venue connected to the event.  Applications operating in this domain that encounter transient network events such as loss or reordering of some packets often experience user-visible artifacts in the media.
+
+Applications requiring ultra low latency for media delivery are usually tightly constrained on the available choices for media transport technologies, and sometimes may need to operate in controlled environments to reliably achieve their latency and quality goals.
+
+Most applications operating over IP networks and requiring latency this low use the real-time transport protocol (RTP) {{RFC3550}} or WebRTC {{I-D.draft-ietf-rtcweb-overview-19}}{{WEBRTC}}, which uses RTP for the media transport as well as several other protocols necessary for safe operation in browsers.
```

Comment:
Is it worth mentioning the continued use of [Adobe RTMP] here, particularly common still as an ingest protocol?

[Adobe RTMP]: https://rtmp.veriskope.com/docs/spec/

Reply:
Yes, probably.  Thanks!

Reply:
If you're adding RTMP, I believe https://tools.ietf.org/html/rfc7425 would be a reasonable reference. 

We came close to dying, getting that through the RFC process (it started out as AD sponsored in the IETF stream), so it seems like we should get some benefit from publishing it!

Reply:
The mention is good, but the wording  doesn't reflect how the situation is still developing.      Is it worth mentioning the different use cases for Ultra LL?      Game watching like Twitch is one use case, while live sports is a different one.

Reply:
Good point, @gitnnelg! 

If we mention specific protocols here, do we also want to describe typical workflow stages?

This section is about "Ultra Low Latency" described in terms of overall glass-to-glass latency, but RTMP (for example) is also still commonly used as an ingest protocol for other categories with higher glass-to-glass latency tolerances.

If we do want to mention workflow stages here, the recent SVA document, ["Best Practices for End-to-End Workflow Monitoring"] may be a useful reference. 

["Best Practices for End-to-End Workflow Monitoring"]: https://www.streamingvideoalliance.org/document/best-practices-for-end-to-end-workflow-monitoring/

Reply:
I like the idea of referencing the SVA doc as it pulls in industry references which anchors the doc to actual industry activity.

Mentioning the workflow stages, at least at a high level, seems good as it links the how to the why/when.

Reply:
IMHO the ingest use case is important. When latency is important, this is the segment where RTMP seems to be the most used nowadays. Beyond RTMP, It s difficult to draw the line though. If we wanted to be more inclusive, SRT is being used (by people with a haivision encoder), RIST is being used, RTSP is being used (especially for surveillance camera), and even raw RTP (AWS kinetic), or webrtc (millicast.com).

Very often the protocol choice seems to be dictated by the existing hardware.

Most of the time, with the exception of RTSP/RTP/webrtc, it includes a transcoding step in the media path which adds latency. The added latency is at least one frame (33ms), but depends on the depth of the frame buffer, and is usually inconsequential unless you are aiming for real-time (gaming, ...) where your total glass to glass latency budget is usually 150~200 ms. A recent peer-reviewed study on comparison of encoders for real-time content touches on that point [preprint](https://arxiv.org/abs/2009.14165)

I could not find an authoritative source on corresponding volumes, even though bitomovin 2020 report put them in that order: HLS > RTMP > webrtc. I believe this makes sense only if you take "slow live" (more than 10s) into account, and there HLS prevails, otherwise the consensus is that RTMP prevails.

To make supporting WebRTC ingest easier, especially for hardware encoders, there is an ongoing IETF effort and a proposed Draft. It's going through Dispatch this time around, so it s too early to put it in this document, but I wanted to bring it to the attention of the group, and it could be of interest for discussion in march, dispatch willing.
https://datatracker.ietf.org/doc/draft-murillo-whip/

Reply:
Great comments in this thread, thanks!

This discussion makes me think maybe the right thing is to open yet another issue about writing a section that outlines workflow stages in a glass to glass pipeline (for which I agree the SVA doc looks like a very good reference).

I think of particular interest for this doc would be a focus on where the transport between stages does or does not commonly happen over IP networks, along with pointers to commonly used technologies and a brief summary of the networking considerations (like the pros and cons, provisioning and reliability requirements, along with common failure modes and their causes).

I think having such a section about workflow stages would help a lot with establishing better context for this latency section.  So I'll leave some time to raise objections (and maybe even ask the wg opinion tomorrow), but otherwise plan to open that issue shortly, unless someone has a better suggestion.

Reply:
> Is it worth mentioning the continued use of [Adobe RTMP](https://rtmp.veriskope.com/docs/spec/) here, particularly common still as an ingest protocol?

I would vote for not including this. RTMP is nearing its death (Adobe is no longer supporting it and content providers are migrating from it), leaving the space to newer approaches like SRT, RIST, Zixi.

Reply:
If the WG decides that we need to discuss ingest in this document, that probably needs its own section. It would not be the "latency" section. So, closing this comment here.